### PR TITLE
Refactor Hash Keys

### PR DESCRIPTION
**Pull request by virena**
- Excluded: `NoOp` and `input` so far
- Separate params into header files (I've left implementations in `.cc` files; there are only a few that are lengthy - `Pool2d`,`Conv2d`, and `Flat`)
- Some validity checks that require `config` in parallel ops are still in `get_or_create_[op]_node()` - is this the best practice? or should we make `config` a param and check in `is_valid()` instead?
@lockshaw

### CODE REVIEW

File: include/flexflow/ops/conv_2d_params.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  bool is_valid(ParallelTensorShape const &input) const;
+  void solve_dims(ParallelTensorShape const &input,
+                  ParallelDim output_dims[MAX_TENSOR_DIM],
+                  int *output_ndims,
+                  ParallelDim kernel_dims[MAX_TENSOR_DIM],
+                  int *kernel_ndims,
+                  ParallelDim bias_dims[MAX_TENSOR_DIM],
+                  int *bias_ndims) const;
+  // size_t get_hash(const ParallelTensor input) const;
```
Comment (lockshaw) [inline include/flexflow/ops/conv_2d_params.h:25]:
```suggestion
```

File: include/flexflow/ops/dropout.h [hunk 2]
```diff
@@ -1,26 +1,32 @@
...
+  using Input = ParallelTensor;
   Dropout(FFModel &model,
           const ParallelTensor input,
           float rate,
           unsigned long long seed,
           char const *name);
   Dropout(FFModel &model, Dropout const &other, const ParallelTensor input);
+  Dropout(FFModel &model,
+          Params const &params,
+          const Input input,
```
Comment (lockshaw) [inline include/flexflow/ops/dropout.h:28]:
```suggestion
          Input const input,
```
for consistency

File: include/flexflow/ops/element_unary_params.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+#include "flexflow/ffconst.h"
+#include "flexflow/parallel_tensor.h"
+
+namespace FlexFlow {
+
+struct ElementUnaryParams {
+  OperatorType op_type;
+  bool inplace;
+  float scalar;
```
Comment (lockshaw) [inline include/flexflow/ops/element_unary_params.h:12]: I don't think all unary ops have a `scalar`--is there a change this will create problems?

File: include/flexflow/ops/linear_params.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+class LinearParams {
+public:
+  LayerID layer_guid;
+  int out_channels;
+  bool use_bias;
+  DataType data_type;
+  ActiMode activation;
+
+  bool is_valid(ParallelTensorShape const &input_shape) const;
+  void solve_dims(const ParallelTensor input,
```
Comment (lockshaw) [inline include/flexflow/ops/linear_params.h:19]: Out of scope, but because `ParallelTensor` is just a `ParallelTensorBase *`, `ParallelTensor const` is actually a `const` pointer to a non-`const` `ParallelTensorBase`, which I'm pretty sure is not the desired behavior (correct me if I'm wrong here @jiazhihao). We should really fix this at some point, probably by defining a proper `ParallelTensor` containing a `std::shared_ptr<ParallelTensorBase>`

File: include/flexflow/ops/split.h [hunk 5]
```diff
@@ -1,17 +1,28 @@
 #ifndef _FLEXFLOW_SPLIT_H
 #define _FLEXFLOW_SPLIT_H
 
-#include "flexflow/model.h"
+#include "flexflow/device.h"
+#include "flexflow/fftype.h"
+#include "flexflow/layer.h"
+#include "flexflow/node.h"
+#include "flexflow/op_meta.h"
+#include "flexflow/operator.h"
```
Comment (lockshaw) [inline include/flexflow/ops/split.h:10]: Do you need to `#include "flexflow/ops/split_params.h`?

File: include/flexflow/ops/split_params.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+#include "flexflow/parallel_tensor.h"
+
+namespace FlexFlow {
+
+struct SplitParams {
+  std::vector<int> split;
+  int legion_axis;
+  bool is_valid(ParallelTensorShape const &) const;
+} bool operator==(SplitParams const &, SplitParams const &);
```
Comment (lockshaw) [inline include/flexflow/ops/split_params.h:12]:
```suggestion
};
bool operator==(SplitParams const &, SplitParams const &);
```

File: include/flexflow/parallel_ops/fused_parallel_op_params.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+};
+bool operator==(FusedParallelOpParams const &, FusedParallelOpParams const &);
+
+} // namespace FlexFlow
+
+namespace std {
+template <>
+struct hash<FlexFlow::FusedParallelOpParams> {
+  size_t operator()(FlexFlow::FusedParallelOpParams const &) const;
+}
```
Comment (lockshaw) [inline include/flexflow/parallel_ops/fused_parallel_op_params.h:18]: Why is there a semicolon after the struct specialization in other files but not here?

File: src/ops/cast.cc [hunk 8]
```diff
@@ -14,6 +14,7 @@
  */
 
 #include "flexflow/ops/cast.h"
+#include "flexflow/model.h"
```
Comment (lockshaw) [inline src/ops/cast.cc:17]: Why is `model.h` included here? Is there something weaker (maybe `cast_params.h`?) that we can use here instead? `model.h` pulls in and contains a lot, so this could potentially trigger a lot of potentially unnecessary rebuilds

File: src/ops/cast.cc [hunk 9]
```diff
@@ -63,26 +64,29 @@ Op *Cast::create_operator_from_layer(
...
-  Cast *cast;
-  auto const &it = this->cached_cast_ops.find(hash);
-  if (it != cached_cast_ops.end()) {
-    cast = it->second;
-  } else {
-    cast = new Cast(*this, input, dtype, NULL);
-    cached_cast_ops[hash] = cast;
-  }
+bool CastParams::is_valid(ParallelTensorShape const &input) const {
+  // TODO: more check on the input shape
```
Comment (lockshaw) [inline src/ops/cast.cc:82]: Are there more checks that should be done here? What are they?

File: src/ops/cast.cc [hunk 10]
```diff
@@ -63,26 +64,29 @@ Op *Cast::create_operator_from_layer(
...
 // }
 
+CastParams Cast::get_params() const {
+  CastParams params;
+  params.dtype = this->dtype;
+  return params;
+}
+
 using PCG::Node;
 Node FFModel::get_or_create_cast_node(const ParallelTensor input,
```
Comment (lockshaw) [inline src/ops/cast.cc:74]: I think all `get_or_create_[op]_node` methods should be discarded. If you see a case where that can't be done, lmk and I can take a look

File: src/ops/attention.cc [hunk 11]
```diff
@@ -36,6 +37,15 @@ using Legion::Task;
 using Legion::TaskArgument;
 using Legion::TaskLauncher;
 
+bool MultiHeadAttentionParams::is_valid(
+    std::vector<ParallelTensorShape> const &input) const {
+  bool is_valid = true;
+  for (auto pt_shape : input) {
```
Comment (lockshaw) [inline src/ops/attention.cc:43]:
```suggestion
  for (auto const &pt_shape : input) {
```

File: src/ops/element_binary.cc [hunk 12]
```diff
@@ -108,9 +108,8 @@ Tensor FFModel::divide(const Tensor in1,
 }
 
 bool ElementBinaryParams::is_valid(
-    std::pair<ParallelTensorShape, ParallelTensorShape> const &) const {
-  // TODO: more check on the input shape
-  return true;
+    std::pair<ParallelTensorShape, ParallelTensorShape> const &input) const {
+  return input.first.is_valid() && input.second.is_valid();
```
Comment (lockshaw) [inline src/ops/element_binary.cc:112]: We probably also want a check that the sizes of the two inputs are the same

File: src/ops/embedding.cc [hunk 13]
```diff
@@ -90,21 +91,18 @@ EmbeddingParams Embedding::get_params() const {
...
 
-size_t EmbeddingParams::get_hash(const ParallelTensor input) const {
-  size_t hash = input->get_owner_independent_hash();
-  hash_combine(hash, this->num_entries);
-  hash_combine(hash, this->out_channels);
-  hash_combine(hash, this->aggr);
-
-  return hash;
+bool EmbeddingParams::is_valid(ParallelTensorShape const &input) const {
+  return input.is_valid();
```
Comment (lockshaw) [inline src/ops/embedding.cc:99]: Should there be a `TODO` here? Is this really the only constraint for `Embedding`?

File: src/ops/embedding.cc [hunk 14]
```diff
@@ -14,6 +14,7 @@
  */
```
Comment (lockshaw) [inline src/ops/embedding.cc:14]: I think git's a bit confused about this file, can you resolve the merge conflicts and then I'll take another look at it?

File: src/ops/attention.cc [hunk 15]
```diff
@@ -844,6 +859,29 @@ bool MultiHeadAttention::measure_operator_cost(
 
 using PCG::Node;
 
+bool operator==(MultiHeadAttentionParams const &lhs,
```
Comment (lockshaw) [inline src/ops/attention.cc:862]: I tried to use `std::tie` to remove the verbosity of code like this and potentially allow you to template-automate a bunch of this code. There's an example at https://github.com/lockshaw/FlexFlow/blob/feat/candle_split/include/flexflow/node.h#L40-L42 and https://github.com/lockshaw/FlexFlow/blob/feat/candle_split/src/runtime/node.cc#L15-L17. Not saying you definitely should use it (I'm on the fence), but you can give it a look and lmk what you think

File: src/ops/reshape.cc [hunk 16]
```diff
@@ -34,7 +35,20 @@ using Legion::Task;
...
 using Legion::TaskLauncher;
 
-ReshapeParams::ReshapeParams(std::vector<int> const &_shape) : shape(_shape) {}
+/* Params */
+bool operator==(ReshapeParams const &lhs, ReshapeParams const &rhs) {
+  return lhs.shape == rhs.shape;
+}
+
+bool ReshapeParams::is_valid(ParallelTensorShape const &input) const {
+  return input.is_valid();
```
Comment (lockshaw) [inline src/ops/reshape.cc:44]: There should probably also at least be a volume check if not some others here against the shape field in `ReshapeParams`

File: src/parallel_ops/reduction.cc [hunk 17]
```diff
@@ -37,6 +38,23 @@ using Legion::Task;
...
 using Legion::TaskLauncher;
 
+/* Params */
+bool operator==(ReductionParams const &lhs, ReductionParams const &rhs) {
+  return lhs.reduction_legion_dim == rhs.reduction_legion_dim &&
+         lhs.reduction_degree == rhs.reduction_degree;
+}
+
+bool ReductionParams::is_valid(ParallelTensorShape const &input) const {
+  return input.is_valid();
```
Comment (lockshaw) [inline src/parallel_ops/reduction.cc:48]: Missing some constraints here

**Review by lockshaw — CHANGES_REQUESTED** [review]

File: include/flexflow/ops/attention.h [hunk 18]
```diff
@@ -1,14 +1,23 @@
...
+#include "flexflow/ops/attention_params.h"
 
 namespace FlexFlow {
 
 class MultiHeadAttentionMeta;
 
 class MultiHeadAttention : public Op {
 public:
+  using Params = MultiHeadAttentionParams;
+  using Input = std::vector<ParallelTensor>;
```
[the remark this answers is not in the document]
Comment (virena) [reply include/flexflow/ops/attention.h:19]: I modeled this one similar to `concat.h` which takes a vector, but you're right I think we can make `attention.h` more specific. `tuple<,,>` makes sense to me.

File: include/flexflow/ops/element_unary_params.h (hunk 3 above)
Comment (virena) [reply include/flexflow/ops/element_unary_params.h:12]: I'll add a default value. I believe this is how it's currently handled in the other constructor based on the implementation of `FFModel::unary`

File: include/flexflow/parallel_ops/fused_parallel_op_params.h (hunk 7 above)
Comment (virena) [reply include/flexflow/parallel_ops/fused_parallel_op_params.h:18]: That's odd. I would've thought that gets caught by the formatter. Just added it to the other parallel ops too.

File: src/ops/cast.cc (hunk 8 above)
Comment (virena) [reply src/ops/cast.cc:17]: I believe this is because of `get_or_create_node` (and in general, `FFModel`). I used [this diff](https://github.com/flexflow/FlexFlow/pull/273/files#diff-1a3382f4659ec99ef9690c3a4158c6a71732dc1b6641d03796bb797ed45fbd0c) for reference. `model.h` is removed from `cast.h` and instead included in `cast.cc`.

Comment (virena) [reply src/ops/cast.cc:17]: It looks like the discussion on that PR also seems to acknowledge how large `model.h` is -- maybe we can separate the node creation stuff? Not sure how to approach this but if `model.h` is what is causing the bulk of the rebuild then separating params may not contribute too much unless we resolve that issue.

File: src/ops/cast.cc (hunk 9 above)
Comment (virena) [reply src/ops/cast.cc:82]: I forgot to delete this comment. See [here](https://github.com/flexflow/FlexFlow/blob/master/src/ops/cast.cc#L66) for reference. I think this is the only check required.

File: src/ops/cast.cc (hunk 10 above)
Comment (virena) [reply src/ops/cast.cc:74]: I actually do think it's possible for all of them. I was initially a bit confused about `config` for the parallel ops, but it looks like in `substitution.cc` we can take advantage of `model.config` in `GraphXfer::create_new_operator` (so the code in the parallel ops was actually redundant).

File: src/ops/element_binary.cc (hunk 12 above)
Comment (virena) [reply src/ops/element_binary.cc:112]: From my understanding of `ParallelTensorShape` this can be done with `==`. Just wanted to confirm that this is the correct way to check if two shapes are the same.

File: src/ops/embedding.cc (hunk 13 above)
Comment (virena) [reply src/ops/embedding.cc:99]: Based on [this](https://github.com/flexflow/FlexFlow/blob/master/src/ops/embedding.cc#L715), it does not seem like there are other constraints.

File: src/ops/attention.cc (hunk 15 above)
Comment (virena) [reply src/ops/attention.cc:862]: Ah, this seems helpful. Will certainly take a look. Found some other issues with the PR while reviewing so will probably get around to this after I fix some of those.

File: src/ops/reshape.cc (hunk 16 above)
Comment (virena) [reply src/ops/reshape.cc:44]: Maybe I am going about the validity checks incorrectly, so this would be helpful to clarify. I've been referring to the [original function](https://github.com/flexflow/FlexFlow/blob/master/src/ops/reshape.cc#L426) for the right checks to make. It seems like there are some `assert` calls in the constructor on the actual `ParallelTensor` input volume. If we are still doing those, then do we need additional volume checks on the shape in node creation?

File: src/parallel_ops/reduction.cc (hunk 17 above)
Comment (virena) [reply src/parallel_ops/reduction.cc:48]: Same comment as above for validity checks.

**Comment by virena** [comment]
Realized there are some more errors in the PR - will fix and then re-request review.

File: src/ops/cast.cc (hunk 8 above)
Comment (lockshaw) [reply src/ops/cast.cc:17]: Ah yeah, I think I mixed up the `.cc` and `.h` files and though I was looking at the other one. The main goal is to cause fewer kernel rebuilds as they are much slower, so as long as only the `.cc` files and not the `.h` files include `model.h`, then the impact on search time shouldn't be too bad. In the long run I think the direction to go in is to pull the kernel logic out of the operator class itself (see <https://github.com/lockshaw/FlexFlow/blob/feat/candle_split/include/flexflow/ops/linear_kernels.h> for an example) so that the operator interface can be refactored freely without ever requiring a kernel rebuild.

File: src/ops/cast.cc (hunk 9 above)
Comment (lockshaw) [reply src/ops/cast.cc:82]: Sounds good, then let's just make sure we remove the comment before we merge. On a side topic, do you have any idea of what the existing check is doing/why it's there?

File: src/ops/element_binary.cc (hunk 12 above)
Comment (lockshaw) [reply src/ops/element_binary.cc:112]: Yeah that should be fine.

File: src/ops/reshape.cc (hunk 16 above)
Comment (lockshaw) [reply src/ops/reshape.cc:44]: `[...]Params::is_valid` is used to determine the return value of `Model::get_or_create_node<T>`, which is then used in `GraphXfer::create_new_operator`, which is called from `Graph::find_matches` to determine if a substitution applys to an input graph. Therefore, if a check is missing from `is_valid`, instead of rejecting that substitution and going and explore other ones, we'll apply that substitution, construct the operator, fail the assertion, and crash out of search (which is bad). Thus the checks in `is_valid` should generally be a superset of the assertions in the operator itself except in cases where a check would indicate that something is more wrong than just an invalid subsetitution being tried out. 

This case probably won't happen too frequently as the substitution generator is in theory supposed to only produce substitutions where the src ops and constraints matching implies that the dst ops and constraints should always work, so this is why I'm guessing this doesn't come up as often as it otherwise would. That said, I think it's probably a good behavior to maintain from now and as the generator -> Unity pipeline gets refined more we can consider potentially changing it. @jiazhihao Correct me if I'm forgetting anything here.

 > I've been referring to the [original function](https://github.com/flexflow/FlexFlow/blob/master/src/ops/reshape.cc#L426) for the right checks to make
 
 Unfortunately the original functions frequently only include a subset of the valid checks.

File: include/flexflow/parallel_ops/fused_parallel_op_params.h (hunk 7 above)
Comment (lockshaw) [reply include/flexflow/parallel_ops/fused_parallel_op_params.h:18]: Is the semicolon required to compile? The formatter's role (at least according to our formatter) generally consists of taking syntactically correct code and producing correctly formatted syntactically correct code, so if the input code is broken it's kinda allowed to do whatever I think, which might(?) explain it? 

As a side note currently there is no validation that the code still builds after a commit, so if you can always validate locally that everything still builds before you merge.

File: src/ops/cast.cc (hunk 8 above)
Comment (virena) [reply src/ops/cast.cc:17]: Yes, that makes sense to me.

File: src/ops/reshape.cc (hunk 16 above)
Comment (virena) [reply src/ops/reshape.cc:44]: That makes sense. So if the original `get_or_create_[op]_node` functions only contained a subset of the valid checks, then we probably need to look at `is_valid` for each operator more closely and see if there are more checks required.

File: src/ops/cast.cc (hunk 9 above)
Comment (virena) [reply src/ops/cast.cc:82]: I'm not sure what the purpose of that check is. Actually, I don't think I could explain most of the checks that deal with `degree` and parallel dims. Before we merge this, it'd help to discuss each operator since some of the checks are insufficient as it is.

**Comment by virena** [comment]
- Removed all `get_or_create_[op]_node` with the exception of `noop`, `input`, `fused_parallel`, and `parallel_op`. The parallel ones have some control flow that dictates which `get_or_create_node<[Op]>` gets called, so I have left them as is.
- Inconsistency in `substitution.cc`: in some cases, `opx->matchOpx->mapOp.ptr` is casted to an operator pointer (see: `Conv2D`). This allows us to call `get_params()`. In other cases, this is not done. 
- Inconsistency in `graph.cc`: some operators have a `deserialize` function in their implementation files and some don't.

For both of these files, I have followed the same pattern as was in the file before, but I am wondering if we should fix these inconsistencies. @lockshaw

File: src/runtime/parallel_op.cc [hunk 19]
```diff
@@ -62,17 +62,25 @@ Node FFModel::get_or_create_parallel_op_node(
 
   switch (op_type) {
     case OP_COMBINE:
-      return this->get_or_create_combine_node(
-          input, parallel_dim, parallel_degree);
+      CombineParams combine_params;
+      combine_params.combine_legion_dim = parallel_dim;
```
Comment (lockshaw) [inline src/runtime/parallel_op.cc:66]:
```suggestion
      combine_params.legion_dim = parallel_dim;
```
Feels like this would be more concise (not a huge deal either way)

File: src/runtime/substitution.cc [hunk 20]
```diff
@@ -700,8 +700,11 @@ void Graph::reshape_output_tensor(ParallelTensorShape const &desired_shape) {
       assert(desired_degree % current_degree == 0);
       int partition_factor = desired_degree / current_degree;
 
-      Node partition_node = model->get_or_create_repartition_node(
-          output_tensor, i, partition_factor);
+      RepartitionParams params;
+      params.repartition_legion_dim = i;
+      params.repartition_degree = partition_factor;
+      Node partition_node =
+          model->get_or_create_node<Repartition>(output_tensor, params);
```
Comment (lockshaw) [inline src/runtime/substitution.cc:707]:
```suggestion
          model->get_or_create_node<Repartition>(output_tensor, {i/*legion_dim*/, partition_factor});
```
Where not ambiguous this might make the code less verbose

File: src/runtime/substitution.cc [hunk 21]
```diff
@@ -841,7 +847,16 @@ bool GraphXfer::create_new_operator(OpX const *opx, Node &op) {
     case OP_CONCAT: {
       int axis;
       assert(opx->get_pm_constraint(PM_AXIS, axis));
-      op = model->get_or_create_concat_node(opx->inputs.size(), inputs, axis);
+
```
Comment (lockshaw) [inline src/runtime/substitution.cc:850]: Why can't we just pass in `inputs` here as opposed to doing a rather verbose copy into `_inputs`?

File: include/flexflow/ops/attention.h [hunk 22]
```diff
@@ -130,6 +130,7 @@ class MultiHeadAttention : public Op {
   bool add_bias_kv, add_zero_attn;
   int qSize, kSize, vSize, qProjSize, kProjSize, vProjSize, oProjSize;
   int qoSeqLength, kvSeqLength;
+  int embed_dim, kdim, vdim;
```
Comment (lockshaw) [inline include/flexflow/ops/attention.h:133]: Why is adding these necessary?

File: src/ops/flat.cc [hunk 23]
```diff
@@ -371,7 +365,8 @@ Node Flat::deserialize(FFModel &ff,
                        ParallelTensor inputs[],
                        int num_inputs) {
   assert(num_inputs == 1);
-  return ff.get_or_create_flat_node(inputs[0]);
+  FlatParams params;
+  return ff.get_or_create_node(inputs[0], params);
```
Comment (lockshaw) [inline src/ops/flat.cc:369]:
```suggestion
  return ff.get_or_create_node<Flat>(inputs[0], {});
```

File: src/ops/reshape.cc (hunk 16 above)
Comment (lockshaw) [reply src/ops/reshape.cc:44]: Just to write down what we discussed in the meeting: except for the graph cycle checks the substitution generator should be responsible for maintaining correctness through substitution application, not the runtime. As such we should eventually remove the `is_valid` checks (#301)

File: src/parallel_ops/combine.cc [hunk 24]
```diff
@@ -53,14 +53,14 @@ bool CombineParams::is_valid(ParallelTensorShape const &input) const {
...
-  params.combine_legion_dim = this->combine_legion_dim;
+  params.combine_legion_dim = this->combine_dim;
   params.combine_degree = this->combine_degree;
   return params;
 }
 
 Combine::Combine(FFModel &model,
                  CombineParams const &params,
-                 const ParallelTensorInput input,
+                 const ParallelTensor input,
```
Comment (lockshaw) [inline src/parallel_ops/combine.cc:63]:
```suggestion
                 ParallelTensor const input,
```

File: src/parallel_ops/fused_parallel_op.cc [hunk 25]
```diff
@@ -281,7 +285,7 @@ size_t hash<FlexFlow::FusedParallelOpParams>::operator()(
     FlexFlow::FusedParallelOpParams const &params) const {
   size_t key = 0;
   hash_combine(key, params.parallel_ops.size());
-  for (ParallelOpInfo const &p : params.parallel_ops) {
+  for (auto const &p : params.parallel_ops) {
```
Comment (lockshaw) [inline src/parallel_ops/fused_parallel_op.cc:288]: Any reason for this change? I generally prefer to use explicit type names over `auto` when they're short just for clarity, but happy to be convinced otherwise :slightly_smiling_face:

File: src/parallel_ops/fused_parallel_op.cc [hunk 26]
```diff
@@ -50,7 +50,11 @@ bool FusedParallelOpParams::is_valid(ParallelTensorShape const &input) const {
 
 FusedParallelOpParams FusedParallelOp::get_params() const {
   FusedParallelOpParams params;
-  params.parallel_ops = this->parallel_ops;
+  std::vector<ParallelOpInfo> vec;
+  for (ParallelOpInfo op : this->parallel_ops) {
+    vec.push_back(op);
+  }
+  params.parallel_ops = vec;
```
Comment (lockshaw) [inline src/parallel_ops/fused_parallel_op.cc:57]:
```suggestion
  params.parallel_ops = this->parallel_ops;
```
I think the additional copy is unnecessary

File: src/parallel_ops/partition.cc [hunk 27]
```diff
@@ -96,7 +96,7 @@ Repartition::Repartition(FFModel &model,
 
 Repartition::Repartition(FFModel &model,
                          RepartitionParams const &params,
-                         const ParallelTensorInput input,
+                         const ParallelTensor input,
```
Comment (lockshaw) [inline src/parallel_ops/partition.cc:99]:
```suggestion
                         ParallelTensor const input,
```

File: src/parallel_ops/reduction.cc [hunk 28]
```diff
@@ -92,7 +92,7 @@ Reduction::Reduction(FFModel &model,
 
 Reduction::Reduction(FFModel &model,
                      ReductionParams const &params,
-                     const ParallelTensorInput input,
+                     const ParallelTensor input,
```
Comment (lockshaw) [inline src/parallel_ops/reduction.cc:95]:
```suggestion
                     ParallelTensor const input,
```

File: src/parallel_ops/replicate.cc [hunk 29]
```diff
@@ -92,7 +92,7 @@ Replicate::Replicate(FFModel &model,
 
 Replicate::Replicate(FFModel &model,
                      ReplicateParams const &params,
-                     const ParallelTensorInput input,
+                     const ParallelTensor input,
```
Comment (lockshaw) [inline src/parallel_ops/replicate.cc:95]:
```suggestion
                     ParallelTensor const input,
```

File: src/parallel_ops/replicate.cc [hunk 30]
```diff
@@ -50,7 +50,7 @@ bool ReplicateParams::is_valid(ParallelTensorShape const &input) const {
 
 ReplicateParams Replicate::get_params() const {
   ReplicateParams params;
-  params.replicate_legion_dim = this->replicate_legion_dim;
```
Comment (lockshaw) [inline src/parallel_ops/replicate.cc]: Why this change?

File: src/runtime/graph.cc [hunk 31]
```diff
@@ -1911,7 +1911,13 @@ void FFModel::deserialize_graph_optimal_view(
       case OP_CONCAT: {
         int legion_axis;
         dez.deserialize(legion_axis);
-        node = get_or_create_concat_node(num_inputs, inputs, legion_axis);
+        std::vector<ParallelTensor> _inputs;
```
Comment (lockshaw) [inline src/runtime/graph.cc:1914]: I don't think this copy is necessary

File: src/runtime/graph.cc [hunk 32]
```diff
@@ -2002,61 +2019,72 @@ void FFModel::deserialize_graph_optimal_view(
...
+        params.embed_dim = embed_dim;
+        params.num_heads = num_heads;
+        params.kdim = k_dim;
+        params.vdim = v_dim;
+        params.dropout = dropout;
+        params.bias = bias;
+        params.add_bias_kv = add_bias_kv;
+        params.add_zero_attn = add_zero_attn;
+        params.layer_guid = layer_guid;
+        node = get_or_create_node<MultiHeadAttention>(_inputs, params);
```
Comment (lockshaw) [inline src/runtime/graph.cc:2034]:
```suggestion
        node = get_or_create_node<MultiHeadAttention>({inputs[0], inputs[1], inputs[2]}, params);
```

File: src/runtime/graph.cc [hunk 33]
```diff
@@ -1948,7 +1962,10 @@ void FFModel::deserialize_graph_optimal_view(
         assert(num_inputs == 2);
         OperatorType op_type;
         dez.deserialize(op_type);
-        node = get_or_create_element_binary_node(inputs[0], inputs[1], op_type);
+        auto input_pair = std::make_pair(inputs[0], inputs[1]);
+        ElementBinaryParams params;
+        params.type = opx->type;
+        node = get_or_create_node<ElementBinary>(input_pair, params);
```
Comment (lockshaw) [inline src/runtime/graph.cc:1968]:
```suggestion
        node = get_or_create_node<ElementBinary>({inputs[0], inputs[1]}, params);
```

File: src/ops/embedding.cc [hunk 34]
```diff
@@ -101,8 +101,8 @@ bool EmbeddingParams::is_valid(ParallelTensorShape const &input) const {
```
Comment (lockshaw) [inline src/ops/embedding.cc:101]: Is there any way you can make this diff more readable? Is there some merge conflict that needs to be addressed?

**Review by lockshaw — CHANGES_REQUESTED** [review]

File: include/flexflow/ops/linear_params.h (hunk 4 above)
Comment (lockshaw) [reply include/flexflow/ops/linear_params.h:19]: Tracking in #302

File: src/ops/cast.cc (hunk 8 above)
Comment (lockshaw) [reply src/ops/cast.cc:17]: Tracking in #303

**Comment by lockshaw** [comment]
> * Inconsistency in `substitution.cc`: in some cases, `opx->matchOpx->mapOp.ptr` is casted to an operator pointer (see: `Conv2D`). This allows us to call `get_params()`. In other cases, this is not done.
 I don't see any issue with using `opx->matchOpX->mapOp.ptr`, but it's not necessary that everything match this pattern if it doesn't make sense, and even if it does you can just add a `TODO` and we can refactor those over later as we go along. I'd rather get this PR merged sooner rather than later :slightly_smiling_face: 

> * Inconsistency in `graph.cc`: some operators have a `deserialize` function in their implementation files and some don't.
Long term I think we'll want to move to auto-generating `serialize` and `deserialize` from the params structs, but for this PR I wouldn't worry about this inconcistency

File: src/runtime/parallel_op.cc (hunk 19 above)
Comment (virena) [reply src/runtime/parallel_op.cc:66]: All the parallel op files are pretty explicit when prefixing `dim` and `degree` so I followed that pattern for now. However, now that I'm thinking about it, could we have a unified `ParallelOpParams` for combine, replicate, repartition, and reduce? They all have the same structure, and `cached_ops` maintains a separate map for each operator anyway.

File: src/runtime/substitution.cc (hunk 20 above)
Comment (virena) [reply src/runtime/substitution.cc:707]: I like that. Made the change in other places where there were at most 2 params.

File: src/runtime/substitution.cc (hunk 21 above)
Comment (virena) [reply src/runtime/substitution.cc:850]: `inputs` is of type `int[]` and we need a `vector<int>` for the template functions that convert `ParallelTensor` to `ParallelTensorShape`. That being said, I've replaced the loop above with `
        std::vector<ParallelTensor> _inputs (std::begin(inputs), std::end(inputs));`  for something more elegant.

File: src/parallel_ops/fused_parallel_op.cc (hunk 25 above)
Comment (virena) [reply src/parallel_ops/fused_parallel_op.cc:288]: I like the ease of `auto` while writing code, but in this instance, I think I was actually just figuring out an issue with my linking as it wasn't picking up on the `ParallelOpInfo` import. I've reverted it back now since it's working.

File: src/parallel_ops/replicate.cc (hunk 30 above)
Comment (virena) [reply src/parallel_ops/replicate.cc]: The original was an error, actually. In `replicate.h`, it is `replicate_dim` and not `replicate_legion_dim`. The nomenclature is actually a little inconsistent, when `[op]_legion_dim` is used vs `[op]_dim`. I've kept the params consistent with the constructor.

File: src/runtime/graph.cc (hunk 31 above)
Comment (virena) [reply src/runtime/graph.cc:1914]: See above comment on `int[]`. I've made the same modification here.

File: src/ops/embedding.cc (hunk 34 above)
Comment (virena) [reply src/ops/embedding.cc:101]: That is strange, I don't see any conflicts on my end. I will take a detailed look and see what the issue is.

File: src/runtime/substitution.cc [hunk 35]
```diff
@@ -841,7 +841,8 @@ bool GraphXfer::create_new_operator(OpX const *opx, Node &op) {
     case OP_CONCAT: {
       int axis;
       assert(opx->get_pm_constraint(PM_AXIS, axis));
-      op = model->get_or_create_concat_node(opx->inputs.size(), inputs, axis);
+      std::vector<ParallelTensor> _inputs(std::begin(inputs), std::end(inputs));
+      op = model->get_or_create_node<Concat>(_inputs, {axis});
```
Comment (lockshaw) [inline src/runtime/substitution.cc:845]:
```suggestion
      op = model->get_or_create_node<Concat>({std::begin(inputs), std::end(inputs)}, {axis});
```
Also, `std::end` works on constant sized arrays? Wasn't aware of this

File: src/runtime/parallel_op.cc (hunk 19 above)
Comment (lockshaw) [reply src/runtime/parallel_op.cc]: I think I'd rather keep each separate parallel op having its own params class for now. I'm happy to be convinced that was should go to a single `ParallelOpParams` but either way let's not make it part of this PR.

File: src/runtime/graph.cc (hunk 31 above)
Comment (lockshaw) [reply src/runtime/graph.cc]: I'd recommend the same change as with the other case

File: src/ops/attention.cc (hunk 15 above)
Comment (lockshaw) [reply src/ops/attention.cc:864]: Let's leave this out of this PR and we can chat sometime about if we want to add it separately

**Comment by lockshaw** [comment]
Can you add a TODO to the `embedding.h` or `embedding.cc` reminding to get rid of `layer_guid` and linking to #304?

Otherwise looks good! Once these last two changes are added I think we're done! Lmk when you're ready for a final pass. :slightly_smiling_face:

**Review by lockshaw — APPROVED** [review]

File: include/flexflow/model.h [hunk 36]
```diff
@@ -246,29 +246,43 @@ class NoOp;
 ParallelConfig get_basic_data_parallel_config(int num_parts, int dims);
 
 class Cast;
+class CastParams;
 class Concat;
```
Comment (lockshaw) [inline include/flexflow/model.h:250]: Hey, why are all of these `Param` forward declarations here? Aren't they taken care of by `#include`ing `operator_params.h`?

Comment (virena) [reply include/flexflow/model.h:250]: I think a few of these were there before so I followed that pattern, but could we put `#include operator_params.h` in `model.cc`? I think that might be faster at compile time.
